inginious/frontend/static/js/common.js

```diff
@@ -77,31 +77,41 @@ function registerCodeEditor(textarea, lang, lines, firstline=1)
 
     var is_single = $(textarea).hasClass('single');
 
+     var keyMappings = {
+        'Ctrl-Enter': function() {
+                                 $('body,html').animate({
+                                   scrollTop: $('#task-submit').offset().top
+                                 }, 'fast');
+                               }
+     }
+
+    if (user_indentation_type["text"] == "tabs") {
+        keyMappings["Tab"] = function(cm) { cm.execCommand("insertSoftTab"); let text = cm.getSearchCursor('    '); while(text.find()){text.replace("\t");}; };
```

Comment:
All the 4-space-length blocks will indeed be replaced by tabs if it is the indentation method the user has selected. It is perfectly intended because of Python not handling mixed indentation of spaces and tabs.

It has nothing to do with ```insertTab``` and ```insertSoftTab```. The difference between those two is that the first simply adds a "\t" and the second completes an unfinished indentation (based on the indentUnit) and replaces it with a "\t".

Reply:
Correction : 
`insertTab` inserts a "\t" character at the place of the cursor to match the indentation given by `tabSize`.
`insertSoftTab` inserts a certain amount of spaces to match the indentation given by `tabSize`.

The problem with `insertTab` is it does not replace the already existing spaces so there is still a mix of spaces and tabs.

Reply:
I have found a better way to avoid replacing all quadruple spaces in the code with tabs. I use `insertTab`, `identLess` and `insertTab` again. This will have the same effect of completing a tab and replacing 4-space blocks with tabs but only the ones used for indentation.

What happens is we complete the tab with `insertTab`, delete it with `indentLess` and recreate it in full with the second `insertTab`.  All the previous indentation spaces are replaced by tabs because of the way `indentLess` works and because we use `indentWithTabs=true`.